pkg/mallochook/mallochook.c

```diff
@@ -0,0 +1,208 @@
+static void *(*mallochook_valloc)(size_t size);
+
+static void *(*mallochook_memalign)(size_t alignment, size_t size);
+static void *(*mallochook_pvalloc)(size_t size);
+
+static mallochook_cnt_t mallochook_heap_inuse;
+static mallochook_cnt_t mallochook_heap_alloc;
+
+static void mallochook_track_alloc(void *ptr) {
+    if (ptr != NULL) {
+        size_t usable = malloc_usable_size(ptr);
+        mallochook_inc(mallochook_heap_inuse, usable);
+        mallochook_inc(mallochook_heap_alloc, usable);
+    }
+}
+
+static void mallochook_track_free(void *ptr) {
+    if (ptr != NULL) {
+        size_t usable = malloc_usable_size(ptr);
+        mallochook_dec(mallochook_heap_inuse, usable);
+    }
+}
+
+static void* mallochook_loadsym(const char *name) {
+    dlerror(); // Clear last error, see dlsym man page
+    void *ptr = dlsym(RTLD_NEXT, name);
+    if (ptr == NULL) {
+        char *err = dlerror();
+        if (err == NULL) {
+            err = "symbol is defined, but null";
+        }
+        // printf calls malloc, but we may have just failed to load one
+        const char *reloc_err = "error patching symbol ";
+        // not much we can do about errors or partial writes here, but this
+        // keeps the -Werror happy.
+        int n = 0;
+        n += write(2, reloc_err, strlen(reloc_err));
+        n += write(2, name, strlen(name));
+        n += write(2, ": ", 2);
+        n += write(2, err, strlen(err));
+        n += write(2, "\n", 1);
+    }
+    return ptr;
+}
+
+// Temporary calloc implementation to use while we are loading symbols using
+// dlsym, which in turn calls calloc. This codepath can handle allocation errors
+// and will use a global static buffer instead. mallochook_init will ensure that
+// all symbols are resolved during process startup when only one thread is
+// running.
+static void *mallochook_calloc_stub(size_t nmemb, size_t size) {
+    return NULL;
+}
+
+static void mallochook_load_all(void) {
+    mallochook_calloc = mallochook_calloc_stub;
+    mallochook_calloc = mallochook_loadsym("calloc");
+
+    mallochook_malloc = mallochook_loadsym("malloc");
+    mallochook_realloc = mallochook_loadsym("realloc");
+    mallochook_reallocarray = mallochook_loadsym("reallocarray");
+    mallochook_free = mallochook_loadsym("free");
+
+    mallochook_posix_memalign = mallochook_loadsym("posix_memalign");
+    mallochook_aligned_alloc = mallochook_loadsym("aligned_alloc");
+    mallochook_valloc = mallochook_loadsym("valloc");
+
+    mallochook_memalign = mallochook_loadsym("memalign");
+    mallochook_pvalloc = mallochook_loadsym("pvalloc");
+}
+
+static void mallochook_ensure_loaded(void) {
+    if (mallochook_calloc == NULL) {
+        mallochook_load_all();
+    }
+}
+
+static void mallochook_init(void) __attribute__((constructor));
+void mallochook_init(void) {
+    mallochook_ensure_loaded();
+}
+
+void *malloc(size_t size) {
+    mallochook_ensure_loaded();
```

Comment:
While constructor will always run, it is possible that it will not run early enough: constructors in shared libraries are run first, and they can call malloc too. In fact, libstdc++ [allocates some global memory](https://github.com/gcc-mirror/gcc/blob/releases/gcc-10.2.0/libstdc%2B%2B-v3/libsupc%2B%2B/eh_alloc.cc#L262) for the exception handling code, so this malloc is consistently called before mallochook_init.